source/Cosmos.HAL2/Drivers/PCI/Audio/AC97.cs

```diff
@@ -0,0 +1,301 @@
+                }
+
+                bufferDescriptorList[i].bufferSize = bufferSize;
+                bufferDescriptorList[i].configuration |= BD_FIRE_INTERRUPT_ON_CLEAR;
+            }
+
+            // Tell BDL location
+            fixed (void* ptr = bufferDescriptorList)
+            {
+                pBufferDescriptors.DWord = (uint)ptr;
+            }
+
+            // Set last valid index
+            lastValidIdx = 2;
+            pLastValidEntry.Byte = lastValidIdx;
+        }
+
+        private void HandleInterrupt(ref INTs.IRQContext aContext)
+        {
+            ushort sr = pTransferStatus.Word;
+
+            if ((sr & IRQ_LVBCI) > 0)
+            {
+                // Last Valid Buffer interrupt
+                pTransferStatus.Word = IRQ_LVBCI;
+            }
+            else if ((sr & IRQ_BCIS) > 0)
+            {
+                // Load a buffer ahead
+                int next = lastValidIdx + 1;
+                if (next >= BUFFER_COUNT)
+                    next -= BUFFER_COUNT;
+
+                BufferProvider.RequestBuffer(transferBuffer);
+
+                fixed (byte* mainBufPtr = transferBuffer.RawData)
+                {
+                    MemoryOperations.Copy(
+                        dest: bufferDescriptorList[next].pointer,
+                        src: mainBufPtr,
+                        size: bufferSizeBytes
+                    );
+                }
+
+                // Set the index to the current one
+                lastValidIdx++;
+                if (lastValidIdx == BUFFER_COUNT)
+                    lastValidIdx = 0;
+
+                pLastValidEntry.Byte = lastValidIdx;
+                pTransferStatus.Word = IRQ_BCIS;
+            }
+            else if ((sr & IRQ_FIFO_ERROR) > 0)
+            {
+                pTransferStatus.Word = IRQ_FIFO_ERROR;
+            }
+        }
+
+        public override SampleFormat[] GetSupportedSampleFormats()
+            => new SampleFormat[]
+            {
+                new SampleFormat(AudioBitDepth.Bits16, 2, true)
+                // TODO: Implement more channels, as the AC97 spec defines 2/4/6 channel audio
+            };
+
+        public override void SetSampleFormat(SampleFormat sampleFormat)
+        {
+            if (sampleFormat.bitDepth != AudioBitDepth.Bits16)
+                throw new NotSupportedException("The AC97 driver only supports 16-bit audio.");
+
+            if (sampleFormat.channels != 2)
+                throw new NotSupportedException("The AC97 driver only supports stereo audio.");
+
+            if (!sampleFormat.signed)
+                throw new NotSupportedException("The AC97 driver does not support unsigned audio.");
+
+            // TODO: The AC97 specification defines support 2/4/6 channel audio. Currently, only stereo audio output is supported.
+        }
+
+        public override bool Enabled =>
+            (pTransferControl.Byte & TC_RUN_OR_PAUSE) != 0;
+
+        public override void Enable()
```

Comment:
I assume enabling an already enabled driver would cause it to reset?

Reply:
Oops, yeah, a call to `Enable()` should be ignored if the driver is already enabled. Sorry, missed that one, will change that!